src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java

```diff
@@ -265,4 +323,171 @@ public void onError(IOException e, HttpURLConnection uc) throws IOException {
+                        )));
+                    } else {
+                        expiration = rateLimit.getResetDate().getTime()
+                                + (ENTROPY.nextInt() & 0xffff);
+                        listener.getLogger().println(GitHubConsoleNote.create(System.currentTimeMillis(), String.format(
+                                "GitHub API Usage: Current quota has %d remaining (%d over budget). Next quota of %d in %s. Sleeping until reset.",
+                                rateLimit.remaining, ideal - rateLimit.remaining, rateLimit.limit,
+                                Util.getTimeSpanString(rateLimitResetMillis)
+                        )));
+                    }
+                } else {
+                    // work out how long until remaining == ideal + 0.1 * buffer (to give some spend)
+                    double targetFraction = (rateLimit.remaining - buffer * 1.1) / (rateLimit.limit - buffer - burst);
+                    expiration = rateLimit.getResetDate().getTime()
+                            - Math.max(0, (long) (targetFraction * MILLIS_PER_HOUR))
+                            + ENTROPY.nextInt(1000);
+                    listener.getLogger().println(GitHubConsoleNote.create(System.currentTimeMillis(), String.format(
+                            "GitHub API Usage: Current quota has %d remaining (%d over budget). Next quota of %d in %s. Sleeping for %s.",
+                            rateLimit.remaining, ideal - rateLimit.remaining, rateLimit.limit,
+                            Util.getTimeSpanString(rateLimitResetMillis),
+                            Util.getTimeSpanString(expiration - System.currentTimeMillis())
+                    )));
+                }
+                long nextNotify = System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(180);
+                while (expiration > System.currentTimeMillis()) {
+                    long wake = Math.min(expiration, nextNotify);
+                    while (wake > System.currentTimeMillis()) {
+                        if (Thread.interrupted()) {
+                            throw new InterruptedException();
+                        }
+                        Thread.sleep(Math.max(1,System.currentTimeMillis() - wake));
+                    }
+                    nextNotify += TimeUnit.SECONDS.toMillis(180);
+                    long now = System.currentTimeMillis();
+                    if (now < expiration) {
+                        GHRateLimit current = github.getRateLimit();
+                        if (current.remaining > rateLimit.remaining
+                                || current.getResetDate().getTime() > rateLimit.getResetDate().getTime()) {
+                            listener.getLogger().println(GitHubConsoleNote.create(now,
+                                    "GitHub API Usage: The quota may have been refreshed earlier than expected, rechecking..."
```

Comment:
True…in this case we probably need to do a sweep of messages used in various plugins (and Jenkins core) which only appear in build logs. I am pretty sure there are a lot of these.

There could be an `I18NConsoleNote` in core which you could use to print something rendered in the browser’s locale. But then the plain-text build log will be hard to read.